src/phylo/iteration.jl

```diff
@@ -0,0 +1,277 @@
+""" -> 
+function Base.start(x::Tip2Root)
+  return (x.start, false)
+end
+
+@Docile.doc """
+Step to the next node when iterating over a `Phylogeny`.
+
+**Parameters:**
+* `x`: A `Tip2Root` iterator.
+* `state`: A `Tuple` that contains the node to be visited and whether the root has been reached. 
+""" ->
+function Base.next(x::Tip2Root, state::Tip2RootState)
+  return state[1], (state[1].parent, isroot(state[1]))
+end
+
+@Docile.doc """
+Check whether the iterator is at the last node in a phylogeny and is therefore
+done iterating.
+
+**Parameters:**
+* `x`: A `Tip2Root` `PhylogenyIterator`.
+* `state`: a `Tuple` containing the `PhyNode` that is to be visited next, and whether the root has been reached.
+""" ->
+function Base.done(x::Tip2Root, state::Tip2RootState)
+  return state[2]
+end
+
+@Docile.doc """
+Begin iteration over an individual PhyNode i.e. all other PhyNodes connected to it,
+so parent and children.
+
+**Parameters:**
+* `x`: A PhyNode to iterate over.
+""" ->
+function Base.start(x::PhyNode)
+  if parentisself(x)
+    return (1, x.children)
+  else
+    return (1, [x.children, x.parent])
+  end
+end
+
+@Docile.doc """
+Step to the next node when iterating over a `Phylogeny`.
+
+**Parameters:**
+* `x`: A `PhyNode`.
+* `state`: A `Tuple` that contains the index of the `Array` of `PhyNodes` to visit and the array itself.
+""" ->
+function Base.next(x::PhyNode, state::Tuple{Int64, Array{PhyNode, 1}})
+  current::PhyNode = state[2][state[1]]
+  return current, (state[1] + 1, state[2])
+end
+
+@Docile.doc """
+Check whether iteration over all connections of a PhyNode is complete.
+
+**Parameters:**
+* `x`: A `PhyNode`.
+* `state`: A `Tuple` that contains the index of the `Array` of `PhyNodes` to visit and the array itself.
+""" ->
+function Base.done(x::PhyNode, state::Tuple{Int64, Array{PhyNode, 1}})
+  return state[1] > length(state[2])
+end
+
+@Docile.doc """
+Iterate over a `Phylogeny` and return a reference to the first `PhyNode` 
+that matches a given condition.
+
+**Parameters:** 
+* `it`: One of any type of `PhylogenyIterator`. This dictates the order in which nodes will be searched.
+* `condition`: A `Function` that accepts a `PhyNode` as the first and only argument, and returns a Bool.
+""" ->
+function Base.search{T <: PhylogenyIterator}(it::T, condition::Function)
+  for i = it
+    # I feel like the function should catch occasions where the provided condition does not return a Bool.
+    if condition(i)
+      return i
+    end
+  end
+end
+
+@Docile.doc """
+Iterate over a `Phylogeny` and return an array of references to all `PhyNode`s 
+that match a given condition.
+
+**Parameters:** 
+* `it`: One of any type of `PhylogenyIterator`. This dictates the order in which nodes will be searched.
+* `condition`: A `Function` that accepts a `PhyNode` as the first and only argument, and returns a Bool.
+""" ->
+function searchall{T <: PhylogenyIterator}(it::T, condition::Function)
```

Comment:
Similarly, `filter(condition, it)` is pretty much the same thing but general purpose.